providers/events.md

```diff
@@ -0,0 +1,257 @@
+
+  private
+
+  # The Stream class isn't a requirement but helps to encapsulate the logic of
+  # fetching events from the provider.
+  def event_monitor_handle
+    @event_monitor_handle ||= begin
+      self.class.parent::Stream.new
+    end
+  end
+end
+```
+
+The Stream implementation is going to be very specific to your provider, this is just an example of how one might look.
+
+`app/models/manageiq/providers/awesome_cloud/cloud_manager/event_catcher/stream.rb`
+```ruby
+class ManageIQ::Providers::AwesomeCloud::CloudManager::EventCatcher::Stream
+  attr_reader :ems, :stop_polling, :poll_sleep
+
+  def initialize(ems, options = {})
+    @ems = ems
+    @stop_polling = false
+    @poll_sleep = options[:poll_sleep] || 20.seconds
+  end
+
+  def start
+    @stop_polling = false
+  end
+
+  def stop
+    @stop_polling = true
+  end
+
+  def poll
+    since = nil
+
+    loop do
+      ems.with_provider_connection(:service => "Events") do |events_client|
+        # Replace with your provider's events API
+        events = events_client.get_events(:since => since)
+        since = Time.now.utc
+
+        break if stop_polling
+
+        events.each { |event| yield event }
+      end
+      sleep(poll_sleep)
+    end
+  end
+end
+```
+
+### Parsing events
+
+Now that we're catching events from the provider we need to implement the parser.  This is called by the main event catcher runner thread prior to putting events onto MiqQueue for the MiqEventHandler to process.
+
+The purpose of this is similar to the inventory parser, to translate the native provider event into the VMDB schema.
+
+`app/models/manageiq/providers/awesome_cloud/cloud_manager/event_parser.rb`
+```ruby
+module ManageIQ::Providers::AwesomeCloud::CloudManager::EventParser
+  def self.event_to_hash(event, ems_id)
+    event_hash = {
+      :event_type => event[:event_name],
+      :source     => "AWESOME_CLOUD",
+      :ems_ref    => event[:id],
+      :timestamp  => event[:time],
+      :full_data  => event,  # Make sure this serializes properly and isn't dumping a native object
+    }
+
+    # Parse any additional information e.g. instances, images, etc...
+    case event_hash[:event_type]
+    when /com.awesome-cloud.vm/
+      parse_vm_event!(event, event_hash)
+    end
+
+    event_hash
+  end
+
+  def self.parse_vm_event!(event, event_hash)
+    return if event[:vm].nil?
+
+    # The uid_ems/ems_ref should match the attributes that you set in the inventory
+    # parser since they will be used to lookup the VM object by the MiqEventHandler
+    event_hash[:vm_uid_ems] = event.dig(:vm, :id)
+    event_hash[:vm_ems_ref] = event.dig(:vm, :id)
+    event_hash[:vm_name]    = event.dig(:vm, :name)
+  end
+end
+```
+
+And with that you should be able to catch events and have them saved to the database!
+
+To test this out you can run `simulate_queue_worker` in a rails console (this will act as your `MiqEventHandler`) and then you can run you event catcher with `run_single_worker`:
```

Comment:
```suggestion
To test this out you can run `simulate_queue_worker` in a rails console (this will act as your `MiqEventHandler`) and then you can run your event catcher with `run_single_worker`:
```

Reply:
:+1: done